File: CacheManager.m
//
//  AvroKeyboard
//
//  Created by Rifat Nabi on 7/1/12.
//  Copyright (c) 2012 OmicronLab. All rights reserved.
//

#import "CacheManager.h"

static CacheManager* sharedInstance = nil;

@implementation CacheManager

+ (CacheManager *)sharedInstance  {
    if (sharedInstance == nil) {
        [[self alloc] init]; // assignment not done here, see allocWithZone
    }
	return sharedInstance;
}

+ (id)allocWithZone:(NSZone *)zone {
    
    if (sharedInstance == nil) {
        sharedInstance = [super allocWithZone:zone];
        return sharedInstance;  // assignment and return on first allocation
    }
    return nil; //on subsequent allocation attempts return nil
}

- (id)copyWithZone:(NSZone *)zone {
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        // Weight PLIST File
        NSString *path = [self getSharedFolder];
        NSFileManager *fileManager = [NSFileManager defaultManager];
        if ([fileManager fileExistsAtPath:path] == NO) {
            NSError* error = nil;
            [[NSFileManager defaultManager] createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:&error];
            if (error) {
                @throw error;
            }
        }
        
        path = [path stringByAppendingPathComponent:@"weight.plist"];
        
        if ([fileManager fileExistsAtPath:path]) {
            _weightCache = [[NSMutableDictionary alloc] initWithContentsOfFile:path];
        } else {
            _weightCache = [[NSMutableDictionary alloc] initWithCapacity:0];
        }
        _phoneticCache = [[NSMutableDictionary alloc] initWithCapacity:0];
        _recentBaseCache = [[NSMutableDictionary alloc] initWithCapacity:0];
    }
    return self;
}

- (void)dealloc {
    [self persist];
}

- (void)persist {
    [_weightCache writeToFile:[[self getSharedFolder] stringByAppendingPathComponent:@"weight.plist"] atomically:YES];
}

- (NSString*)getSharedFolder {
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
    return [[paths[0] 
             stringByAppendingPathComponent:@"OmicronLab"] 
            stringByAppendingPathComponent:@"Avro Keyboard"];
}

// Weight Cache
- (NSString*)stringForKey:(NSString*)aKey {
    return _weightCache[aKey];
}

- (void)removeStringForKey:(NSString*)aKey {
    [_weightCache removeObjectForKey:aKey];
}

- (void)setString:(NSString*)aString forKey:(NSString*)aKey {
    _weightCache[aKey] = aString;
}

// Phonetic Cache
- (NSArray*)arrayForKey:(NSString*)aKey {
    return _phoneticCache[aKey];
}

- (void)setArray:(NSArray*)anArray forKey:(NSString*)aKey {
    _phoneticCache[aKey] = anArray;
}

// Base Cache
- (void)removeAllBase {
    [_recentBaseCache removeAllObjects];
}

- (NSArray*)baseForKey:(NSString*)aKey {
    return _recentBaseCache[aKey];
}

- (void)setBase:(NSArray*)aBase forKey:(NSString*)aKey {
    _recentBaseCache[aKey] = aBase;
}

@end

File: CacheManager.swift
//
//  CacheManager.swift
//  AvroKeyboard
//
//  Created by Max on 4/25/17.
//
//

import Foundation


@objc public class CacheManager: NSObject {
    fileprivate(set) var weightCache:[String:String] = [:]
    fileprivate(set) var phoneticCache:[String:Array<Any>] = [:]
    fileprivate(set) var recentBaseCache:[String:Array<Any>] = [:]
    
    public static let shared = CacheManager()
    
    public func persist() {
        (weightCache as NSDictionary).write(toFile: filePath, atomically: true)
    }
    
    var sharedFolder: String {
        let paths = NSSearchPathForDirectoriesInDomains(.applicationSupportDirectory, .userDomainMask, true)
        return ((paths[0] as NSString).appendingPathComponent("OmicronLab") as NSString).appendingPathComponent("Avro Keyboard")
    }
    
    var filePath: String { return  (sharedFolder as NSString).appendingPathComponent("weight.plist") }
    
    public override init() {
        super.init()
        
        if FileManager.default.fileExists(atPath: filePath),
            let dictionary = NSDictionary(contentsOfFile: filePath) as? [String : String] {
            weightCache = dictionary
        } else if FileManager.default.fileExists(atPath: sharedFolder) == false {
            //io should be done on background thread
            DispatchQueue.global(priority: .background).async {
                try? FileManager.default.createDirectory(atPath: self.sharedFolder, withIntermediateDirectories: true, attributes: nil)
            }
        }
    }
    
    deinit {
        persist()
    }
}

// TODO - Rewrite the CacheManager with meaningful methods

// Weight Cache (default for String)
public extension CacheManager {
    public func string(forKey key:String) -> String? {
        return weightCache[key]
    }
    
    public func removeString(forKey key:String) {
        weightCache[key] = nil
    }
    
    public func setString(_ string:String, forKey key:String) {
        weightCache[key] = string
    }
}

// Phonetic Cahce (default for Array)
public extension CacheManager {
    public func array(forKey key:String) -> Array<Any>? {
        return phoneticCache[key]
    }
    
    public func setArray(_ array:Array<Any>, forKey key:String) {
        phoneticCache[key] = array
    }
}


// Base Cahce
public extension CacheManager {
    public func removeAllBase() {
        recentBaseCache.removeAll()
    }
    
    public func base(forKey key:String) -> Array<Any>? {
        return recentBaseCache[key]
    }
    
    public func setBase(_ base:Array<Any>, forKey key:String) {
        recentBaseCache[key] = base
    }
}
